go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go

```diff
@@ -804,6 +811,85 @@ func TestSchemaChange(t *testing.T) {
+		createv1 := fmt.Sprintf("create view %s as select id from %s", v1name, v2name)
+
+		sql := fmt.Sprintf("%s; %s;", createv2, createv1)
+		var vuuids []string
+		t.Run("create two views, expect both complete", func(t *testing.T) {
+			uuidList := testOnlineDDLStatement(t, createParams(sql, ddlStrategy+" --allow-concurrent --in-order-completion", "vtctl", "", "", true)) // skip wait
+			vuuids = strings.Split(uuidList, "\n")
+			assert.Equal(t, 2, len(vuuids))
+			for _, uuid := range vuuids {
+				onlineddl.WaitForMigrationStatus(t, &vtParams, shards, uuid, normalWaitTime, schema.OnlineDDLStatusComplete, schema.OnlineDDLStatusFailed)
+			}
+		})
+		require.Equal(t, 2, len(vuuids))
+		testTableCompletionTimes(t, vuuids[0], vuuids[1])
+	})
+	t.Run("in-order-completion: new table column, new view depends on said column", func(t *testing.T) {
+		// The VIEW creation can only succeed when the ALTER has completed and the table has the new column
+		t1uuid = testOnlineDDLStatement(t, createParams(alterExtraColumn, ddlStrategy+" --allow-concurrent --postpone-completion --in-order-completion", "vtctl", "", "", true))                // skip wait
+		v1uuid := testOnlineDDLStatement(t, createParams(createViewDependsOnExtraColumn, ddlStrategy+" --allow-concurrent --postpone-completion --in-order-completion", "vtctl", "", "", true)) // skip wait
+
+		testAllowConcurrent(t, "t1", t1uuid, 1)
+		testAllowConcurrent(t, "v1", v1uuid, 1)
+		t.Run("expect table running, expect view ready", func(t *testing.T) {
+			onlineddl.WaitForMigrationStatus(t, &vtParams, shards, t1uuid, normalWaitTime, schema.OnlineDDLStatusRunning)
+			onlineddl.WaitForMigrationStatus(t, &vtParams, shards, v1uuid, normalWaitTime, schema.OnlineDDLStatusQueued, schema.OnlineDDLStatusReady)
+			time.Sleep(ensureStateNotChangedTime)
+			// nothing should change
+			onlineddl.WaitForMigrationStatus(t, &vtParams, shards, t1uuid, normalWaitTime, schema.OnlineDDLStatusRunning)
+			onlineddl.WaitForMigrationStatus(t, &vtParams, shards, v1uuid, normalWaitTime, schema.OnlineDDLStatusQueued, schema.OnlineDDLStatusReady)
+		})
+		t.Run("complete both", func(t *testing.T) {
+			onlineddl.CheckCompleteAllMigrations(t, &vtParams, len(shards)*2)
+		})
+		t.Run("expect table success", func(t *testing.T) {
+			status := onlineddl.WaitForMigrationStatus(t, &vtParams, shards, t1uuid, normalWaitTime, schema.OnlineDDLStatusComplete, schema.OnlineDDLStatusFailed)
+			fmt.Printf("# Migration status (for debug purposes): <%s>\n", status)
+			onlineddl.CheckMigrationStatus(t, &vtParams, shards, t1uuid, schema.OnlineDDLStatusComplete)
+		})
+		t.Run("expect view success", func(t *testing.T) {
+			status := onlineddl.WaitForMigrationStatus(t, &vtParams, shards, v1uuid, normalWaitTime, schema.OnlineDDLStatusComplete, schema.OnlineDDLStatusFailed)
+			fmt.Printf("# Migration status (for debug purposes): <%s>\n", status)
+			onlineddl.CheckMigrationStatus(t, &vtParams, shards, v1uuid, schema.OnlineDDLStatusComplete)
+		})
```

Comment:
Same comment here. The dependency here can be used to confirm order but I think only if we require that they both actually complete.

Reply:
We do validate that they both `complete`.  `"expect table success"` validates that the table's migration is `complete`, and `"expect view success"` validates the the view's migration is complete.
The above does not validate order. Order is validated later on.

Reply:
Yes, you're right. Here we do `onlineddl.CheckMigrationStatus(t, &vtParams, shards, t1uuid, schema.OnlineDDLStatusComplete)` which returns an error, w/o waiting, if the current status does not match. Sorry about that.

Side note:
This logic still doesn't make much sense to me as I don't see how that is better/different than returning an error in the wait if we don't get completed status by the timeout. In my mind the wait should return an error if the status you're waiting for doesn't occur within the provided timeout. Having this split into different calls invites flakiness by the author not knowing or forgetting that wait doesn't actually wait / return an error. Technically you don't even *need* to pass Failed in that call other than to stop waiting sooner. 

I'm sorry for being a PITA but we really need to get the flakiness of these under control. It's been a year or so now with it being a constant issue so I'm being hyper/over critical here (I know that it's unpleasant, for both of us). I think we could eliminate lines of code, simplify the tests, and make them more robust for all of us with: 
```
onlineddl.WaitForMigrationStatus(t, &vtParams, shards, v1uuid, normalWaitTime, schema.OnlineDDLStatusComplete)
```

And that having a require/assert and printing the message if we don't get what we want in the provided timeout:
```
"Expected migration status of Completed not seen for UUID <uuid> within the timeout of 60s. Last seen status: Failed\n"
```

This would make the discussion above moot because we'd clearly fail the test immediately if we didn't reach that Completed status and not have to wonder what happens after (since we're testing that they all succeed and complete in order).  I'll put that on myself though for some refactors later on to clean up some of the logic where we wait for different things then we check and it's not always clear why e.g. So banking this for myself later and will resolve this thread.

Reply:
> This logic still doesn't make much sense to me as I don't see how that is better/different than returning an error in the wait if we don't get completed status by the timeout.

I believe I have a good answer for that! The problem this is trying to solve is excessive CI runtime. Yes, we _could_ just wait for `complete` state. We can wait for `30sec` (or whatever it is), and if by that time we still don't see `complete`, then the test has failed.

But with GitHub CI, and with so many tests in this suite, this can accumulate and get us to tests timeouts. So, we take advantage of the following fact: if the test is `failed`, then that's a terminal state. There is no way for it to magically transition into `success`! And so if it is `failed`, we can stop waiting. That can shave off dozens of seconds. It actually does! Since there's some tests we _expect_ to fail.

And so the general idea is this:

- we run a migration
- we wait until it has reached a terminal state: either `complete` or `failed`; we know there's no point waiting beyond any of these
- we then validate what the terminal state actually was. Typically we validate that it is `complete`

Does that make more sense?

Some other tests, BTW, wait until the state is `queued` or `ready`. This happens with some of the "postponed" tests. Since these states are non-terminal, there _could be_ a scenario where a migration will then transition into `running` or `complete`. In some of these cases we do an extra sleep after reaching `queued` or `ready`, and then _again_ check to see that the state is still in `queued` or `ready`.

Reply:
Yeah, that does make sense! I would then prefer to have the wait always check for failed status and if it sees that then fail the test via require/assert. We don't need to figure this out here, just something to consider for later.

Reply:
Except for those cases where we _expect_ the migration to fail...